models/badges.js

```diff
@@ -27,36 +27,106 @@ const fetchBadges = async ({ size = 100, page = 0 }) => {
-        if (item.data()?.users?.includes(userID)) {
-          const { title, description } = item.data();
-          userBadges.push({ title, description });
-        }
-      });
+    if (!result.userExists) {
+      return { userExists: false, badges };
     }
-
-    return { userExists, userBadges };
+    const userId = result.user.id;
+    const badgeIdsSnapshot = await userBadgeModel.where("userId", "==", userId).get();
+    const badgeDocReferences = badgeIdsSnapshot.docs.map((doc) => {
+      const badgeId = doc.get("badgeId");
+      return firestore.doc(`badges/${badgeId}`);
+    });
+    // INFO: getAll accepts unpacked array
+    const badgesSnapshot = await firestore.getAll(...badgeDocReferences);
+    badges = badgesSnapshot.map((doc) => convertFirebaseDocumentToBadgeDocument(doc.id, doc.data()));
+    return { userExists: true, badges };
   } catch (err) {
     logger.error("Error retrieving user badges", err);
     return err;
   }
-};
+}
+
+/**
+ * Add badge to firestore
+ * @param  badgeInfo { Object }: has badge name, description, imageUrl and createdBy
+ * @return {Promise<{id: string, createdAt: {date: string, time: string}}|Object>}
+ */
+async function createBadge(badgeInfo) {
+  try {
+    const createdAt = admin.firestore.Timestamp.now();
+    // INFO: check if description is missing
+    const description = badgeInfo.description ?? "";
+    const docRef = await badgeModel.add({
+      ...badgeInfo,
+      description,
+      createdAt,
+    });
+    const { date, time } = convertFirebaseTimestampToDateTime(createdAt);
+    const snapshot = await docRef.get();
+    const data = snapshot.data();
+    return { id: docRef.id, ...data, createdAt: { date, time } };
+  } catch (err) {
+    logger.error("Error creating badge", err);
+    return err;
+  }
+}
+
+/**
+ * assign badges to user
+ * @param { Object }: userId: string and badgeIds: Array<string>
+ * @return {Promise<{docIds: Array<string>}|Object>}
+ */
+async function assignBadges({ userId, badgeIds }) {
+  try {
+    const docIds = [];
+    const batch = firestore.batch();
+    badgeIds.forEach((badgeId) => {
+      const ref = userBadgeModel.doc();
+      const id = ref.id;
+      batch.create(ref, { userId, badgeId });
+      docIds.push(id);
+    });
+    await batch.commit();
+    return { docIds };
+  } catch (err) {
+    logger.error("Error assigning badges", err);
+    return err;
+  }
+}
+
+/**
+ * unassign badges from user
+ * @param { Object }: userId: string and badgeIds: Array<string>
+ * @return {Promise<{docIds: Array<string>}|Object>}
+ */
+async function unAssignBadges({ userId, badgeIds }) {
+  try {
+    const docIds = [];
+    const snapshot = await userBadgeModel.where("userId", "==", userId).where("badgeId", "in", badgeIds).get();
+    const batch = firestore.batch();
+    snapshot.forEach((doc) => {
+      docIds.push(doc.id);
+      batch.delete(doc.ref);
+    });
+    await batch.commit();
+    return { docIds };
```

Comment:
The batch operation has a limit of 500 documents. we should add it in logger as well if we find more than 500 docs are there to update.

Reply:
Thanks, for pointing this out. Using functions `chunks` and `assignUnassignBadgesInBulk` writes will happen in chunks of 500 documents per batch.

`example:`
```js
 document-size is 2000
 chunks-size is 500
 totoal-unqiue-batches will be 4
 ```
 
 [Function: assignUnassignBadgesInBulk](https://github.com/Real-Dev-Squad/website-backend/pull/832/commits/d935f6802a95e7f2d9ee71cd84957e0c0548d0af)
 [Function: chunks](https://github.com/Real-Dev-Squad/website-backend/pull/832/commits/e8dcb542649a5cabbb8392c523cd5f9521ef6252)